DatadogRUM/Sources/SDKMetrics/SessionEndedMetric.swift

```diff
@@ -0,0 +1,251 @@
+import DatadogInternal
+
+/// An object tracking the state of RUM session and exporting attributes for "RUM Session Ended" telemetry.
+internal final class SessionEndedMetric {
+    /// Definition of fields in "RUM Session Ended" telemetry, following the "RUM Session Ended" telemetry spec.
+    internal enum Constants {
+        /// The name of this metric, included in telemetry log.
+        /// Note: the "[Mobile Metric]" prefix is added when sending this telemetry in RUM.
+        static let name = "RUM Session Ended"
+        /// Metric type value.
+        static let typeValue = "rum session ended"
+        /// Namespace for bundling metric attributes ("rse" = "RUM Session Ended").
+        static let rseKey = "rse"
+    }
+
+    /// An ID of the session being tracked through this metric object.
+    let sessionID: String
+
+    /// The type of OS component where the session was tracked.
+    private let bundleType: BundleType
+
+    /// The session precondition that led to the creation of this session.
+    private let precondition: RUMSessionPrecondition?
+
+    private struct TrackedViewInfo {
+        let viewURL: String
+        let startMs: Int64
+        var durationNs: Int64
+
+        // TODO: RUM-4591 Track diagnostic attributes:
+        // - `instrumentationType`: manual | uikit | swiftui
+    }
+
+    /// Stores information about tracked views, referencing them by their view ID.
+    @ReadWriteLock
+    private var trackedViews: [String: TrackedViewInfo] = [:]
+
+    /// Info about the first tracked view.
+    @ReadWriteLock
+    private var firstTrackedView: TrackedViewInfo?
+
+    /// Info about the last tracked view.
+    @ReadWriteLock
+    private var lastTrackedView: TrackedViewInfo?
+
+    /// Tracks the number of SDK errors by their kind.
+    @ReadWriteLock
+    private var trackedSDKErrors: [String: Int] = [:]
+
+    /// Indicates if the session was stopped through `stopSession()` API.
+    @ReadWriteLock
+    private var wasStopped = false
+
+    // TODO: RUM-4591 Track diagnostic attributes:
+    // - no_view_events_count
+    // - has_background_events_tracking_enabled
+    // - has_replay
+    // - ntp_offset
+
+    // MARK: - Tracking Metric State
+
+    /// Initializer.
+    /// - Parameters:
+    ///   - sessionID: An ID of the session that is being tracked with this metric.
+    ///   - precondition: The precondition that led to starting this session.
+    ///   - context: The SDK context at the moment of starting this session.
+    init(
+        sessionID: String,
+        precondition: RUMSessionPrecondition?,
+        context: DatadogContext
+    ) {
+        self.sessionID = sessionID
+        self.bundleType = context.applicationBundleType
+        self.precondition = precondition
+    }
+
+    /// Tracks the view event that occurred during the session.
+    func track(view: RUMViewEvent) {
+        guard view.session.id == sessionID else {
+            return // sanity check, unexpected
```

Comment:
maybe worth telemetry?

Reply:
Indeed 👍. I added an error so we will see if views are tracked in wrong sessions (although unexpected, it is possible from upstream code).